### PR TITLE
Fix sympy.testing.pytest.raises to behave like pytest.raises

### PR DESCRIPTION
**Pull request by eric-wieser**
~~There seems to be a mistaken impression that `XFAIL` is needed for tests with `raises`. This should not be true.~~

This was true...

Code written as `assert raises(...)` would pass in pytest and fail in sympy, because sympy was not returning an appropriate object from `raises`.

#### Release Notes

* other
   * `assert sympy.testing.pytest.raises(Exception, func)` no longer always asserts when pytest is not present

### CODE REVIEW

File: sympy/testing/pytest.py [hunk 1]
```diff
@@ -54,6 +63,7 @@ def raises(expectedException, code=None):
         >>> from sympy.testing.pytest import raises
 
         >>> raises(ZeroDivisionError, lambda: 1/0)
+        <ExceptionInfo ZeroDivisionError('division by zero')>
```
Comment (eric-wieser) [inline sympy/testing/pytest.py:66]: Maybe this will fix Python 3.5 CI:
```suggestion
        <ExceptionInfo ZeroDivisionError(...)>
```

**Comment by eric-wieser** [comment]
CI failure looks like a PyPy network issue.

**Comment by asmeurer** [comment]
Looks like it is working after a restart.

**Comment by eric-wieser** [comment]
Good to merge then?

**Comment by oscarbenjamin** [comment]
This looks good and should be merged for compatibility between pytest and sympy's test runner. Personally I think that we should just use pytest all the time but that's another issue.

I also think that we shouldn't use `assert` with `raises`. There are only a handful of cases but they should all be changed. My preferred form is `with raises` but I understand that `raises/lambda` is used for a single liner. Using `assert raises` just seems like a misunderstanding to me.

**Comment by asmeurer** [comment]
I agree `assert raises` might be better if it raised an error (`raises` could return an object with a `__bool__` that gives an error). If we want to be compatible with pytest we should allow it, since pytest apparently doesn't have any issues with `assert raises()`.

I personally like the lambda form better because it avoids this issue

```py
# Test that these all raise ValueError
with raises(ValueError):
    f(...)
    g(...)
    ...
```

That can't work. Only the first exception would be tested. It's an easy mistake to make, even if you know about it. So I would prefer the lambda form unless the thing being tested can only be written as a statement.

**Comment by oscarbenjamin** [comment]
> Only the first exception would be tested.

I agree that this is a problem. Our contributors are generally not sufficiently experienced to avoid that so it is better to prefer the lambda approach most of the time.

Related to that is the `warns` context manager which only works as a context manager. You can see examples where it is being used incorrectly:
```
$ git grep -A 2 'with warns'
...
--
sympy/crypto/tests/test_crypto.py:    with warns(NonInvertibleCipherWarning):
sympy/crypto/tests/test_crypto.py-        assert rsa_public_key(2, 2, 1) == (4, 1)
sympy/crypto/tests/test_crypto.py-        assert rsa_public_key(8, 8, 8) is False
...
```

**Comment by eric-wieser** [comment]
> Personally I think that we should just use pytest all the time but that's another issue.
> 
> I also think that we shouldn't use `assert` with `raises`.

Agreed on both points. If you feel strongly about the second, you could make a patch to pytest to make `ExceptionInfo.__bool__` emit a warning.

**Comment by asmeurer** [comment]
FWIW I'm not a huge fan of pytest. But it's unrelated to this PR, so should be discussed elsewhere.

I don't know if there's any way to have a code quality test that checks against the context manager problem. I think it's kind of like the problem of forgetting to write `assert` on a test, where we just have to keep an eye out in review, and grep the tests every once in a while to fix them.

**Comment by oscarbenjamin** [comment]
I think that this is just something we should pick up on review. It could be listed in the dev guide somewhere as well with an explanation. For now though we can clean it up. There is also an example with `raises`:
```
$ git grep -A 2 'with raises'
...
sympy/matrices/expressions/tests/test_hadamard.py:    with raises(ShapeError):
sympy/matrices/expressions/tests/test_hadamard.py-        hadamard_product(A, C)
sympy/matrices/expressions/tests/test_hadamard.py-        hadamard_product(A, I)
...
```

I don't see any automated way to pick up on the context manager problem without false positives unless we require only a single statement in the block. You could have something like
```
with warns(...):
    x = setup1()
    y = setup2()
    assert test_deprecated(x, y) == 1
```
If the intention is to test that `test_deprecated` works and also gives a warning then this is a valid way of writing the test. We could however require that all setup be moved out of the with statement like:
```
x = setup1()
y = setup2()
with warns(...):
    assert test_deprecated(x, y) == 1
```
That's a more accurate test of the warning but it could lead to less readable tests in some cases.

**Comment by eric-wieser** [comment]
Putting this in to unblock #19463. While the discussion is useful and worth continuing, I don't think it's relevant to whether this PR should be merged.

**Comment by asmeurer** [comment]
I started https://github.com/sympy/sympy/wiki/Writing-tests, with a gotchas section that has these things. I also added some other basic info on writing tests, but more could be added.

**Pull request merged**